search-mapper/src/main/java/org/infinispan/search/mapper/session/impl/InfinispanSearchSession.java

```diff
@@ -0,0 +1,171 @@
+import org.hibernate.search.engine.search.query.dsl.SearchQuerySelectStep;
+import org.hibernate.search.mapper.pojo.model.spi.PojoRuntimeIntrospector;
+import org.hibernate.search.mapper.pojo.session.spi.AbstractPojoSearchSession;
+import org.hibernate.search.mapper.pojo.work.spi.PojoIndexer;
+import org.hibernate.search.util.common.AssertionFailure;
+import org.hibernate.search.util.common.impl.Futures;
+import org.infinispan.search.mapper.common.EntityReference;
+import org.infinispan.search.mapper.common.impl.EntityReferenceImpl;
+import org.infinispan.search.mapper.mapping.EntityConverter;
+import org.infinispan.search.mapper.scope.SearchScope;
+import org.infinispan.search.mapper.scope.impl.SearchScopeImpl;
+import org.infinispan.search.mapper.session.SearchSession;
+import org.infinispan.search.mapper.session.SearchSessionBuilder;
+import org.infinispan.search.mapper.work.SearchIndexingPlan;
+import org.infinispan.search.mapper.work.impl.SearchIndexingPlanImpl;
+
+public class InfinispanSearchSession extends AbstractPojoSearchSession implements SearchSession,
+      DocumentReferenceConverter<EntityReference>, EntityReferenceFactory<EntityReference> {
+
+   private final InfinispanSearchSessionMappingContext mappingContext;
+   private final InfinispanSearchSessionTypeContextProvider typeContextProvider;
+   private final EntityConverter entityConverter;
+
+   private final DocumentCommitStrategy commitStrategy;
+   private final DocumentRefreshStrategy refreshStrategy;
+   private SearchIndexingPlanImpl indexingPlan;
+
+   private InfinispanSearchSession(JavaBeanSearchSessionBuilder builder) {
+      super(builder.mappingContext);
+      this.mappingContext = builder.mappingContext;
+      this.typeContextProvider = builder.typeContextProvider;
+      this.entityConverter = builder.entityConverter;
+      this.commitStrategy = builder.commitStrategy;
+      this.refreshStrategy = builder.refreshStrategy;
+   }
+
+   @Override
+   public void close() {
+      if (indexingPlan != null) {
+         CompletableFuture<?> future = indexingPlan.execute();
+         Futures.unwrappedExceptionJoin(future);
+      }
+   }
+
+   @Override
+   public <E> SearchQuerySelectStep<?, EntityReference, E, ?, ?, ?> search(Collection<? extends Class<E>> types) {
+      return search(scope(types));
+   }
+
+   @Override
+   public <E> SearchQuerySelectStep<?, EntityReference, E, ?, ?, ?> search(SearchScope<E> scope) {
+      return search((SearchScopeImpl<E>) scope);
+   }
+
+   public <E> SearchScopeImpl<E> scope(Collection<? extends Class<? extends E>> types) {
+      return mappingContext.createScope(types);
+   }
+
+   @Override
+   public <T> SearchScope<T> scope(Class<T> expectedSuperType, Collection<String> entityNames) {
+      return mappingContext.createScope(expectedSuperType, entityNames);
+   }
+
+   @Override
+   public SearchIndexingPlan indexingPlan() {
+      if (indexingPlan == null) {
```

Comment:
The session instance is not supposed to be thread safe at the moment, since we create a new instance of session for each request that is bound on a single thread. Sure if we change the way we do indexing, we will have to change that. I think that the change related do this comment could be part of the same [follow up](https://github.com/infinispan/infinispan/pull/8226#discussion_r414439343).